### PR TITLE
[python-package] Implement __eq__ for CatBoost* classes.

### PR DESCRIPTION
**Pull request by daskol**
WIP

### CODE REVIEW

File: catboost/python-package/catboost/core.py [hunk 1]
```diff
@@ -816,6 +816,20 @@ def __deepcopy__(self, _):
         model.__setstate__(state)
         return model
 
+    def __eq__(self, other):
+        if not isinstance(other, _CatBoostBase):
```
Comment (andrey-khropov) [inline catboost/python-package/catboost/core.py]: Предлагаю чтобы не было  copy paste сделать внутреннюю функцию _is_comparable_to(self, rhs) куда поместить общие проверки параметров.

Comment (daskol) [reply catboost/python-package/catboost/core.py]: Updated.

File: catboost/python-package/catboost/core.py [hunk 2]
```diff
@@ -816,6 +816,20 @@ def __deepcopy__(self, _):
...
+            return False
+        if not self.is_fitted() or not other.is_fitted():
```
[the remark this answers is not in the document]
Comment (daskol) [reply catboost/python-package/catboost/core.py]: Done.

**Comment by andrey-khropov** [comment]
Looks good to me but test should be added.

File: catboost/python-package/ut/medium/test.py [hunk 3]
```diff
@@ -3664,3 +3665,40 @@ def test_eval_set_with_no_target_with_eval_metric(task_type):
     )
     with pytest.raises(CatboostError):
         model.fit(train_pool, eval_set=eval_set_pool)
+
+
+def test_model_comparison():
+    def fit_model(value):
+        train_df = titanic()[0].fillna(value)
```
Comment (andrey-khropov) [inline catboost/python-package/ut/medium/test.py]: It's better to use a specific small dataset here that is commited right into repository and used in other tests. Like https://github.com/catboost/catboost/tree/master/catboost/pytest/data/adult. See here for example - https://github.com/catboost/catboost/blob/1aa0ba06cceef82cd4dc64cad084b293ae143839/catboost/python-package/ut/medium/test.py#L196

File: catboost/python-package/ut/medium/test.py [hunk 4]
```diff
@@ -3664,3 +3664,37 @@ def test_eval_set_with_no_target_with_eval_metric(task_type):
...
+        model = CatBoostClassifier(iterations=5)
+        model.fit(pool)
+        return model
+
+    model0 = CatBoostClassifier(iterations=5)
+    model1 = fit_model(42)
+    model2 = fit_model(-999)
+
+    # Test checks that model is fitted.
+    try:
```
Comment (andrey-khropov) [inline catboost/python-package/ut/medium/test.py]: canonical way for pytest is 

```py
with pytest.raises(CatboostError):
    ...
```

like that: https://github.com/catboost/catboost/blob/1aa0ba06cceef82cd4dc64cad084b293ae143839/catboost/python-package/ut/medium/test.py#L539

**Pull request merged**